osf/models/files.py

```diff
@@ -772,6 +805,46 @@ def location_hash(self):
     def archive(self):
         return self.metadata.get('archive')
 
+    def _purge(self, client=None, save=True):
+        if not client:
+            logger.warn(f'No GCS Client detected. Not purging FV {self.id}')
+            return 0
+        if self.basefilenode_set.filter(deleted__isnull=True).exists():
+            logger.warn(f'Live file detected. Not purging FV {self.id}')
+            return 0
+        if not self.location or not self.location.get('object'):
+            logger.warn(f'No valid location detected. Not purging FV {self.id}')
+            return 0
+        dup = FileVersion.objects.exclude(id=self.id).filter(location__object=self.location['object'], basefilenode__deleted__isnull=True).first()
+        if dup:
+            logger.warn(f'Duplicate live file detected on FV {dup.id}. Not purging FV {self.id}')
+            return 0
+        bucket = client.get_bucket(self.location['bucket'])
```

Comment:
is it safe to assume that every file version's `location` has a `'bucket'`?

i fell down [a rabbit hole](https://www.notion.so/cos/FileVersion-location-quest-98b41b78090a4881b0c29a34f1decfbc) trying to answer that question and best i could figure is "probably?" (depends whether the `Region` had it in its `waterbutler_settings` when the `FileVersion` was created, unsure whether old data has been tidied appropriately)

Reply:
For instances where `BaseFileNode.provider` is `osfstorage`, 99% yes. I ran into a similar issue with addon files that have `FileVersion` objects, which is what the above conditional, `if not self.location or not self.location.get('object')` is intended to exclude. Also the `provider='osfstorage'` in the script.